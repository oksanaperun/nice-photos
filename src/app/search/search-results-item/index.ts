export * from './search-results-item.component';
