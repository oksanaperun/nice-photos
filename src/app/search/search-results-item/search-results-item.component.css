.loaded {
  opacity: 1;
  transition: opacity 0.6s ease;
}
