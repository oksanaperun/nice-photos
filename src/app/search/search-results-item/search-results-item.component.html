<img [src]="item.smallUrl"
     [class.hidden]="!isLoaded"
     (load)="onLoaded()"
     (error)="onLoadingFailed()">

<span *ngIf="isLoading" i18n>Loading...</span>
<span *ngIf="isLoadingFailed" i18n>Photo can't be loaded</span>
