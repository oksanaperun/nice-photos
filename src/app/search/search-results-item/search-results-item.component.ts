import { Component, Input } from '@angular/core';
import { Photo } from '../search';

@Component({
  selector: 'app-search-results-item',
  templateUrl: './search-results-item.component.html',
  styleUrls: ['./search-results-item.component.css'],
})

export class SearchResultsItemComponent {
  @Input() item: Photo;

  isLoaded = false;
  isLoadingFailed = false;

  loadingFailedPlaceholder = 'assets/loading_failed.png';

  get itemBoxStyle() {
    if (this.item && (!this.isLoaded || this.isLoadingFailed))
      return {
        'background-color': this.item.color
      };
  }

  get itemStyle() {
    if (this.item && !this.isLoaded)
      return {
        'height': this.item.height + 'px',
        'width': this.item.width + 'px',
        'opacity': 0
      };
  }

  onLoaded(): void {
    this.isLoaded = true;
  }

  onLoadingFailed(): void {
    this.isLoadingFailed = true;
    this.item.smallUrl = this.loadingFailedPlaceholder;
  }
}
