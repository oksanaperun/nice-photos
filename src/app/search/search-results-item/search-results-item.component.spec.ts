import { TestBed, ComponentFixture, async } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { SearchResultsItemComponent } from './search-results-item.component';

describe('SearchResultsItemComponent', () => {
  const photoWithValidImage = { smallUrl: 'data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==' };
  const photoWithInvalidImage = { smallUrl: 'data:image/jpeg;base64,' };
  let component: SearchResultsItemComponent;
  let fixture: ComponentFixture<SearchResultsItemComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [SearchResultsItemComponent]
    });

    fixture = TestBed.createComponent(SearchResultsItemComponent);
    component = fixture.componentInstance;
  });

  it('should show loading text when image is not loaded yet', () => {
    component.item = photoWithValidImage;
    fixture.detectChanges();

    const debugElem = fixture.debugElement.query(By.css('span'));

    expect(debugElem.nativeElement.innerText).toBe('Loading...');
  });

  it('should hide span element when image has loaded', async(() => {
    component.item = photoWithValidImage;
    fixture.detectChanges();

    const debugElem = fixture.debugElement.query(By.css('img'));
    const elem = debugElem.nativeElement;

    elem.onload = () => {
      fixture.detectChanges();

      const debugElems = fixture.debugElement.queryAll(By.css('span'));

      expect(debugElems.length).toBe(0);
    };
  }));

  it('should hide image when image can not be loaded', async(() => {
    component.item = photoWithInvalidImage;
    fixture.detectChanges();

    const debugElem = fixture.debugElement.query(By.css('img'));
    const elem = debugElem.nativeElement;

    elem.onerror = () => {
      fixture.detectChanges();

      expect(elem.style.display).toBe('none');
    };
  }));

  it('should show loading error text when image can not be loaded', async(() => {
    component.item = photoWithInvalidImage;
    fixture.detectChanges();

    const debugElem = fixture.debugElement.query(By.css('img'));
    const elem = debugElem.nativeElement;

    elem.onerror = () => {
      fixture.detectChanges();

      const debugElem = fixture.debugElement.query(By.css('span'));

      expect(debugElem.nativeElement.innerText).toBe('Photo can\'t be loaded');
    };
  }));
});
