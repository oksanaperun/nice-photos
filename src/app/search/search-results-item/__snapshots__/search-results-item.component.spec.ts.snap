// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsItemComponent should notify that image is loading 1`] = `
<app-search-results-item
  isLoaded="false"
  isLoadingFailed="false"
  item={[Function Object]}
>
  <img
    class="hidden"
    src="some-url"
  />
  
  <span>
    Loading...
  </span>
  
  
</app-search-results-item>
`;

exports[`SearchResultsItemComponent should notify that image loading is failed 1`] = `
<app-search-results-item
  isLoaded="false"
  isLoadingFailed={[Function Boolean]}
  item={[Function Object]}
>
  <img
    class="hidden"
    src="some-url"
  />
  
  
  <span>
    Photo can't be loaded
  </span>
  
</app-search-results-item>
`;

exports[`SearchResultsItemComponent should render image 1`] = `
<app-search-results-item
  isLoaded={[Function Boolean]}
  isLoadingFailed="false"
  item={[Function Object]}
>
  <img
    class=""
    src="some-url"
  />
  
  
  
  
</app-search-results-item>
`;
